added popcorntime package

bobvanderlinden opened the pull request:
This adds popcorntime to nixpkgs. Popcorntime is based on node-webkit and is distributed as a .tar.xz, which includes:
- `package.nw`: a zip including all html/css/js
- `Popcorn-Time`: a concatonation of the `nw` binary and `package.nw`
-  `libffmpegsumo.so`: library from Chrome containing the right codecs
- `nw.pak`: a required file related to node-webkit

`Popcorn-Time`, `nw.pak` and `libffmpegsumo.so` must be located in the same directory so that node-webkit will use these files. This is why I used a wrapper script.

The `Popcorn-Time` binary does not run on NixOS, since it links to different library/paths. I build a new `Popcorn-Time` binary from the `nw` binary of NixOS.

I realize using binaries like this isn't the best way to go. I have tried to build `package.nw` through nix, but this took quite a bit of work and reworking the build system (bower + grunt + npm). It eventually didn't result in a correct output (runtime javascript errors).
The work on this can be found here: https://github.com/bobvanderlinden/nixpkgs/tree/popcorntime-source

I've also tried to use `libffmpegsumo.so` of node-webkit in nixpkgs (`${node_webkit_0_9}/share/node-webkit/libffmpegsumo.so`), but it somehow isn't compatible with Popcorntime. In this case the video player does not load correctly. It should be possible to build this file through nix, but I saw this also wasn't the case for node-webkit package. Let me know what is right here.

---

Binary-based version: https://github.com/bobvanderlinden/nixpkgs/tree/popcorntime-binary
Source-based version: https://github.com/bobvanderlinden/nixpkgs/tree/popcorntime-source

bobvanderlinden committed:
added popcorntime package

bobvanderlinden committed:
popcorntime: added support for i686-linux

bobvanderlinden committed:
popcorntime: allow different versions of node_webkit

domenkozar commented:
`pkgs/applications/editors/zed/default.nix` uses node-webkit for inspiration

offlinehacker commented:
Maybe patchelf

V ned., 21. jun. 2015 9:36 dop. je oseba Domen Kožar <
notifications@github.com> napisala:

> pkgs/applications/editors/zed/default.nix uses node-webkit for inspiration
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/8423#issuecomment-113871029>.
>

bobvanderlinden commented:
@domenkozar @offlinehacker Thank you both. I've reworked the non-binary package to generate and fetch the css/js files. This was part of the problem, I've added 2 workarounds, one for grunt and one for bower. I found that @offlinehacker already submitted an issue for the bower problem (https://github.com/bower/bower/issues/1192).

Additionally, there was a problem with playing videos. It seems the `libffmpegsumo.so` that node-webkit comes with doesn't play the videos (missing codecs?). The library (from node-webkit) is not build from source in node-webkit, so I couldn't add the required codecs. Instead I have added the library from PopcornTime's distribution package. To circumvent this we need to build node-webkit and libffmpegsumo.so from source, but this seems like a challenge for a different PR.

Lastly, if this is alright, I can squash the commits into one for a cleaner history.

bjornfor commented:
nix says the the hash for "backbone" is wrong (the one in ```.../popcorntime/bower.nix```).

offlinehacker commented:
My personal opinion would be, to not build from source if that's an
option(Use prebuilt binaries). While we have tools like npm2nix and
bower2nix, I guess in cases like popcorntime, it only makes maintainance of
the package harder, for no real extra benefit.

V pon., 22. jun. 2015 7:41 dop. je oseba Bjørn Forsman <
notifications@github.com> napisala:

> nix says the the hash for "backbone" is wrong (the one in
> .../popcorntime/bower.nix).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/8423#issuecomment-114011147>.
>

bobvanderlinden commented:
@bjornfor Ok, that is a bit scary. I don't know how I could've build everything with that previous hash yesterday... and it's only backbone that's changed. I would almost think the original file of bower has changed :(( (something up with fetchbower?). In any case, I've updated the hash.

@offlinehacker It is indeed much more convenient to maintain while things are working fine without hacks. Updating popcorntime will be a bit clumsy at the moment: updating bower.nix, updating node.nix, possibly incorporating additions to their build process. The result is still a big executable with the zip and nw concatenated and `libffmpegsumo.so` sitting alongside the executable. nw.js is mostly just a big static executable.

In terms of customization of the package and possible security concerns it is still preferable to build from source. However at the moment `nw.js` and `libffmpegsumo.so` are already binary blobs and I would imagine customization and security concerns to be a more significant for nw.js than popcorntime itself (only JS/HTML/CSS).

So, personally I also prefer the binary one. Once `libffmpegsumo.so` and `nw.js` are build from source, it would make a bit more sense to build this from source as well. Even in that case I'm a bit hesitant, since most of it is just JS/CSS/HTML.

I've split this branch into a source-based and a binary-based one. The source-based package can be found here: https://github.com/bobvanderlinden/nixpkgs/tree/popcorntime-source
The binary-based package can be found here: https://github.com/bobvanderlinden/nixpkgs/tree/popcorntime-binary
This PR will include whichever of the two is preferred eventually (source atm).

bjornfor commented:
My experience, regarding binary/source:

I wrote a nix expression for (prebuilt) Popcorn-Time just a few days/weeks ago. I was close to submitting it but then I thought I'd better make it build from source. Building from source was pain (I outright failed), I put it on hold. In the meantime, an update to some library broke my binary Popcorn-Time install (upon re-install of course). So I have mixed feelings about binary/source. I'm OK with what works best, whatever that is... :-)

bobvanderlinden commented:
@bjornfor I started with the source, went to binary and after that finished the source. It might be useful to know the issues I ran into when packaging the source version:

Grunt wanted its 'grunt' packages in the local node_modules. I'm not sure why this requirement is there (probably precaution for version differences) and I couldn't find a nice way to allow the global packages to be used. So I made a workaround for that (make an environment of grunt, link node_modules of the environment locally).

Then there were problems with bower giving errors for not being able to write to its locally installed packages (`.bower.js` specifically). This is because bower copies the packages from the bower.nix environment and also copies the permissions. The permissions are read-only for everyone (including owner). So after that, bower couldn't write its own copied files. (https://github.com/bower/bower/issues/1192)

Now I got PopcornTime to build with a directory of files and a nw wrapper. It somehow gave Javascript errors though.  Apparently there was a problem with airplay-js, which used dependencies that weren't defined in its package.json, but were defined in PopcornTime's package.json. This is only a problem when each package is in its own directory, like NixOS has, but isn't a problem when packages are bundled into a local node_modules. I opened a PR for this issue here: https://git.popcorntime.io/mirrors/node-airplay-js/merge_requests/1. Hopefully it'll be merged.

Then there were Javascript errors because bower2nix missed a few dependencies. I think bower2nix doesn't retrieve packages recursively, but I'm not 100% sure. I still need to figure this one out, but since bower2nix isn't used in nixpkgs, I don't know if it's worth the effort.

Lastly I had trouble with the video player not showing up due to the missing libffmpegsumo.so in combination with not concatenating nw and the zipped directory into one file.

So yea, the binary isn't a bad way to go. Building from source was quite a pain indeed. Future updates could give trouble as well.

offlinehacker commented:
I'm sometimes concerned with purely deterministic aspect of nix, I would
like to have some packages like this to have an option to fetch sources non
deterministically. Making all these hacks to build package is not really
nice and rather makes everything more fragile.

V pon., 22. jun. 2015 9:49 pop. je oseba Bob van der Linden <
notifications@github.com> napisala:

> @bjornfor <https://github.com/bjornfor> I started with the source, went
> to binary and after that finished the source. It might be useful to know
> the issues I ran into when packaging the source version:
>
> Grunt wanted its 'grunt' packages in the local node_modules. I'm not sure
> why this requirement is there (probably precaution for version differences)
> and I couldn't find a nice way to allow the global packages to be used. So
> I made a workaround for that (make an environment of grunt, link
> node_modules of the environment locally).
>
> Then there were problems with bower giving errors for not being able to
> write to its locally installed packages (.bower.js specifically). This is
> because bower copies the packages from the bower.nix environment and also
> copies the permissions. The permissions are read-only for everyone
> (including owner). So after that, bower couldn't write its own copied
> files. (bower/bower#1192 <https://github.com/bower/bower/issues/1192>)
>
> Now I got PopcornTime to build with a directory of files and a nw wrapper.
> It somehow gave Javascript errors though. Apparently there was a problem
> with airplay-js, which used dependencies that weren't defined in its
> package.json, but were defined in PopcornTime's package.json. This is only
> a problem when each package is in its own directory, like NixOS has, but
> isn't a problem when packages are bundled into a local node_modules. I
> opened a PR for this issue here:
> https://git.popcorntime.io/mirrors/node-airplay-js/merge_requests/1.
> Hopefully it'll be merged.
>
> Then there were Javascript errors because bower2nix missed a few
> dependencies. I think bower2nix doesn't retrieve packages recursively, but
> I'm not 100% sure. I still need to figure this one out, but since bower2nix
> isn't used in nixpkgs, I don't know if it's worth the effort.
>
> Lastly I had trouble with the video player not showing up due to the
> missing libffmpegsumo.so in combination with not concatenating nw and the
> zipped directory into one file.
>
> So yea, the binary isn't a bad way to go. Building from source was quite a
> pain indeed. Future updates could give trouble as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/8423#issuecomment-114236529>.
>

bobvanderlinden commented:
Well, before finding bower2nix, I experimented with using bower inside a
builder (like fetchgit etc use). You can just place a sha256 next to it to
be sure you've retrieved what you expected. It's then up to bower to always
retrieve the same files. This could be really handy, not just for bower.

The downside is that it becomes one big package of files (from multiple
bower packages). Not reusable nor efficient when done by many people
multiple times, but for packages like these, yes, it doesn't really matter
anyway.

On Mon, Jun 22, 2015, 22:59 Jaka Hudoklin <notifications@github.com> wrote:

> I'm sometimes concerned with purely deterministic aspect of nix, I would
> like to have some packages like this to have an option to fetch sources non
> deterministically. Making all these hacks to build package is not really
> nice and rather makes everything more fragile.
>
> V pon., 22. jun. 2015 9:49 pop. je oseba Bob van der Linden <
> notifications@github.com> napisala:
>
> > @bjornfor <https://github.com/bjornfor> I started with the source, went
> > to binary and after that finished the source. It might be useful to know
> > the issues I ran into when packaging the source version:
> >
> > Grunt wanted its 'grunt' packages in the local node_modules. I'm not sure
> > why this requirement is there (probably precaution for version
> differences)
> > and I couldn't find a nice way to allow the global packages to be used.
> So
> > I made a workaround for that (make an environment of grunt, link
> > node_modules of the environment locally).
> >
> > Then there were problems with bower giving errors for not being able to
> > write to its locally installed packages (.bower.js specifically). This is
> > because bower copies the packages from the bower.nix environment and also
> > copies the permissions. The permissions are read-only for everyone
> > (including owner). So after that, bower couldn't write its own copied
> > files. (bower/bower#1192 <https://github.com/bower/bower/issues/1192>)
> >
> > Now I got PopcornTime to build with a directory of files and a nw
> wrapper.
> > It somehow gave Javascript errors though. Apparently there was a problem
> > with airplay-js, which used dependencies that weren't defined in its
> > package.json, but were defined in PopcornTime's package.json. This is
> only
> > a problem when each package is in its own directory, like NixOS has, but
> > isn't a problem when packages are bundled into a local node_modules. I
> > opened a PR for this issue here:
> > https://git.popcorntime.io/mirrors/node-airplay-js/merge_requests/1.
> > Hopefully it'll be merged.
> >
> > Then there were Javascript errors because bower2nix missed a few
> > dependencies. I think bower2nix doesn't retrieve packages recursively,
> but
> > I'm not 100% sure. I still need to figure this one out, but since
> bower2nix
> > isn't used in nixpkgs, I don't know if it's worth the effort.
> >
> > Lastly I had trouble with the video player not showing up due to the
> > missing libffmpegsumo.so in combination with not concatenating nw and the
> > zipped directory into one file.
> >
> > So yea, the binary isn't a bad way to go. Building from source was quite
> a
> > pain indeed. Future updates could give trouble as well.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/NixOS/nixpkgs/pull/8423#issuecomment-114236529>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/8423#issuecomment-114260712>.
>

bjornfor commented:
@offlinehacker: Same here. Or even broader, I sometimes question linux packaging ecosystem in general :-)

@bobvanderlinden: Interesting. That sounds similar to what I did with a grails app at work. See fetchGrailsAppDeps in this example: https://gist.github.com/bjornfor/3f2fda1b1ce077225df5#file-grails-app-nix. Yeah, it doesn't re-use build dependencies (it downloads it for each app), but who cares. I will gladly let machines do more work if it means I can do less.

offlinehacker commented:
The problem with npm and bower is that it can fetch sources from non
deterministic locations like latest git commit, and this is quite common.
While npm2nix handles this, what npm2nix does not handle are monthly
changes to npm and new(badly documented) supported formats of dependencies.
In such cases I would just like to have an option to disable this strict
nix determinism(and it's not nix problem, but s*itty package managers like
npm).

V pon., 22. jun. 2015 11:51 pop. je oseba Bjørn Forsman <
notifications@github.com> napisala:

> @offlinehacker <https://github.com/offlinehacker>: Same here. Or even
> broader, I sometimes question linux packaging ecosystem in general :-)
>
> @bobvanderlinden <https://github.com/bobvanderlinden>: Interesting. That
> sounds similar to what I did with a grails app at work. See
> fetchGrailsAppDeps in this example:
> https://gist.github.com/bjornfor/3f2fda1b1ce077225df5#file-grails-app-nix.
> Yeah, it doesn't re-use build dependencies (it downloads it for each app),
> but who cares. I will gladly let machines do more work if it means I can do
> less.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/8423#issuecomment-114274795>.
>

domenkozar commented on pkgs/applications/video/popcorntime/default.nix:
```diff
@@ -0,0 +1,144 @@
+let
+  version = "0.3.7.2";
+  # When upgrading node.nix / node packages:
+  #   fetch package.json from Zed's repository
```
s/Zed/popcorntime/

bobvanderlinden commented:
@domenkozar Good catch. I've fixed the comment in the popcorntime-source branch, but have reset this PR to the popcorn-binary branch. There are 3 votes for using the binary package. What is your opinion?

@offlinehacker Running npm install on dynamic versions goes against the philosophy of Nix to make things reproducible I guess. What would be really nice is if npm could export all of its to-be-installed packages and versions and we can use that export, combined with package.json to install all the static versions. One hash for the whole node_modules is enough in that case..... still though, that's only for npm, bower would need a similar feature... are there more package managers like these?

Anyway, this is exactly what we've talked about before, so I'll reference https://github.com/NixOS/npm2nix/issues/25 for completeness' sake ;)

bjornfor commented on pkgs/applications/video/popcorntime/default.nix:
```diff
@@ -0,0 +1,43 @@
+{ stdenv, pkgs, fetchurl, runCommand, makeWrapper, node_webkit ? pkgs.node_webkit_0_9
```
It is in general considered bad style to pull in the whole `pkgs` set.

bobvanderlinden committed:
popcorntime: don't use pkgs in argument

bobvanderlinden commented:
@bjornfor I've removed the pkgs argument and only added/used node_webkit_0_9.

domenkozar commented:
@bobvanderlinden is this ready to be merged?

bobvanderlinden commented:
@domenkozar The consensus seems to be that the current (binary) version is the way to go. So yes, I think it's ready to merge.

domenkozar commented:
Tested locally, it works well.

bobvanderlinden commented:
Thanks!

bjornfor commented on pkgs/applications/video/popcorntime/default.nix:
```diff
@@ -0,0 +1,43 @@
+
+  popcorntimePackage = stdenv.mkDerivation rec {
+    name = "popcorntime-package-${version}";
+    src = srcs."${stdenv.system}";
```
This line breaks the tarball job:

```
$ nix-build pkgs/top-level/release.nix -A tarball
...
checking pkgs/top-level/all-packages.nix on x86_64-darwin
error: while querying the derivation named ‘popcorntime-0.3.7.2’:
while evaluating the attribute ‘buildCommand’ of the derivation ‘popcorntime-0.3.7.2’ at /tmp/nix-build-nixpkgs-tarball-15.06pre1234.abcdef.drv-0/nixpkgs/pkgs/build-support/trivial-builders.nix:10:14:
while evaluating the attribute ‘src’ of the derivation ‘popcorntime-package-0.3.7.2’ at /tmp/nix-build-nixpkgs-tarball-15.06pre1234.abcdef.drv-0/nixpkgs/pkgs/applications/video/popcorntime/default.nix:19:5:
attribute ‘x86_64-darwin’ missing, at /tmp/nix-build-nixpkgs-tarball-15.06pre1234.abcdef.drv-0/nixpkgs/pkgs/applications/video/popcorntime/default.nix:20:11
build time elapsed:  0m0.047s 0m0.027s 0m40.146s 0m2.255s
builder for ‘/nix/store/g2hidn68pnsyj082fpssgc76jvln1l8h-nixpkgs-tarball-15.06pre1234.abcdef.drv’ failed with exit code 1
error: build of ‘/nix/store/g2hidn68pnsyj082fpssgc76jvln1l8h-nixpkgs-tarball-15.06pre1234.abcdef.drv’ failed
```

bjornfor replied:
Fixed in 8f9534ccc3f3c73a8e17a035cfe059830c679251. (Maybe there is a more elegant way, but it worked.)
